### PR TITLE
Add HTTP/2, QUIC, VPN and Self-Signed CA support

### LINKED ISSUE #463: App does not support custom CAs
**Describe the bug**
I have a private booru instance that I run on my local network.
Said booru instance uses a self-signed certificate that is in a hierarchy of Root CA -> Intermediate CA -> Website CA.
I've installed booth the Root and Intermediate CAs on my phone's trust store.
Everything works fine from browser apps, however this app doesn't seem to use the certificate store from Android (or at least recognize the user-installed certificates).

**To Reproduce**
1. Have a booru instance with a private CA.
2. Add it to the app.
3. App doesn't recognize it.

**Expected behavior**
Booru works normally as the CAs have been installed into the phone and should be recognized as valid.

**Screenshots**
Image showing that the app doesn't manage to make the request to the booru instance successfully

**Desktop (please complete the following information):**
 - OS: N/A

**Smartphone (please complete the following information):**
 - Device: Moto G54 5G
 - OS: Android 14

**Additional context**
Nothing else to add.

### PR DESCRIPTION
**Pull request by GGG-KILLER**
This PR basically makes it so that all requests go through [`package:cronet_http`](https://pub.dev/packages/cronet_http) (on Android) or [`package:cupertino_http`](https://pub.dev/packages/cupertino_http) (on MacOS/iOS) using [`package:native_dio_adapter`](https://pub.dev/packages/native_dio_adapter).

From the [`package:native_dio_adapter`](https://pub.dev/packages/native_dio_adapter) page:

> Using the native platform implementation, rather than the socket-based [dart:io HttpClient](https://api.dart.dev/stable/dart-io/HttpClient-class.html) implementation, has several advantages:
> - It automatically supports platform features such VPNs and HTTP proxies.
> - It supports many more configuration options such as only allowing access through WiFi and blocking cookies.
> - It supports more HTTP features such as HTTP/3 and custom redirect handling.

This enables us in the future to add support for blocking image loading outside of WiFi or metered connections, however I did not want to do more than needed in this PR.

This PR also changes the security settings for the app to allow user-added/user-installed CAs so that people can access their own self-hosted boorus that use self-signed certificates.

Fixes #463.

### CODE REVIEW

File: lib/core/images/dio_extended_image_provider.dart
Comment (GGG-KILLER) [inline lib/core/images/dio_extended_image_provider.dart:1]: Unfortunately there's no other way around other than just vendoring this entire library, wish they'd just offer an extensibility point to `@override` but we'd need to override `_loadNetwork`, and we can't.

File: lib/foundation/networking/dio.dart [hunk 1]
```diff
@@ -51,7 +53,10 @@ Dio newDio(
     },
   ));
 
-  if (supportsHttp2) {
+  // NativeAdapter only does something on Android and iOS/MacOS
+  if (Platform.isAndroid || Platform.isIOS || Platform.isMacOS) {
+    dio.httpClientAdapter = newNativeAdapter();
```
Comment (GGG-KILLER) [inline lib/foundation/networking/dio.dart:58]: There *shouldn't* be any issues to just using the native methods since it's *technically* what the browsers use, so we'd just be doing requests in the same way a browser does, however it'd be nice to have some more testing on this front.

File: lib/foundation/networking/dio.dart [hunk 2]
```diff
@@ -36,7 +37,8 @@ Dio newDio(
   final baseUrl = args.baseUrl;
   final booruFactory = args.booruFactory;
 
-  final booru = booruFactory.getBooruFromUrl(baseUrl);
+  final booru = booruFactory.getBooruFromUrl(baseUrl) ??
+      booruFactory.getBooruFromId(booruConfig.booruId);
```
Comment (GGG-KILLER) [inline lib/foundation/networking/dio.dart:41]: This is needed because otherwise no booru is selected when coming from the "add new booru" page and an unknown URL is used, even though the user has told which booru it is.

File: lib/foundation/networking/dio.dart (hunk 1 above)
Comment (GGG-KILLER) [reply lib/foundation/networking/dio.dart:58]: I've tested gelbooru, gelbooru_v2, shimmie2, anime-pictures and danbooru.

They're all working on Android.

File: lib/foundation/networking/dio.dart [hunk 3]
```diff
@@ -88,6 +93,25 @@ Dio newDio(
...
 
+NativeAdapter newNativeAdapter() {
+  return NativeAdapter(
+    createCronetEngine: () => CronetEngine.build(
+      // We have our own cache interceptor
+      cacheMode: CacheMode.disabled,
+      enableBrotli: true,
+      enableHttp2: true,
+      enableQuic: true,
+    ),
```
Comment (khoadng) [inline lib/foundation/networking/dio.dart:104]: need to add our custom User-Agent

File: lib/foundation/networking/dio.dart (hunk 2 above)
Comment (khoadng) [reply lib/foundation/networking/dio.dart:41]: It works fine on my end. Are you sure this is an issue?

Comment (GGG-KILLER) [reply lib/foundation/networking/dio.dart:41]: Yes, this is an issue when using boorus that don't have known URLs, in my case I have `https://booru.shiro.lan/`, that I tried to add as Danbooru, and when it tries to test it, `booru` is `null` because it tries to get the booru type from the URL, not from what was selected in the dropdown.

Comment (khoadng) [reply lib/foundation/networking/dio.dart:41]: Please merge the latest changes from dev and see if the issue is still there.

Comment (GGG-KILLER) [reply lib/foundation/networking/dio.dart:41]: The issue doesn't manifest anymore because I removed the initial change that only used the native adapter on boorus that had their protocol set as `native`.

It might not cause any issues now since the native adapter is used for everything, but I'd like to prevent this from causing any issues in the future if possible.

File: lib/foundation/networking/dio.dart (hunk 3 above)
Comment (GGG-KILLER) [reply lib/foundation/networking/dio.dart:107]: Done.

File: lib/core/images/dio_extended_image_provider.dart
[the remark this answers is not in the document]
Comment (GGG-KILLER) [reply lib/core/images/dio_extended_image_provider.dart:1]: As mentioned in discord, I've tried to reproduce this to no avail so I'm assuming it was some other unrelated issue, the new logging should enable users to report troubles like this to us should it arise again though.

**Comment by khoadng** [comment]
Looks good, I'll test on iOS/MacOS this weekend.

File: lib/foundation/http/dio_logger_interceptor.dart [hunk 4]
```diff
@@ -27,12 +26,13 @@ class LoggingInterceptor extends Interceptor {
 
   @override
   void onRequest(RequestOptions options, RequestInterceptorHandler handler) {
-    // Don't log image requests
-    final ext = extension(options.uri.toString());
-    if (_kImageExtensions.contains(ext)) {
-      super.onRequest(options, handler);
-      return;
-    }
+    // NOTE: Commented out while new image request system is still new
```
Comment (khoadng) [inline lib/foundation/http/dio_logger_interceptor.dart:29]: Logging all image requests seems unnecessary, maybe we could just log failed ones selectively

Comment (khoadng) [reply lib/foundation/http/dio_logger_interceptor.dart:29]: I'll fix it myself, need to merge this so another refactoring can be done.
